implement Server Timing

i110 opened the pull request:
[Server Timing](https://www.w3.org/TR/server-timing/) allows the clients to get fine-grained performance metrics from the server. This PR only implement `total:dur=xxx` metrics as a trailer.

TODOs:
- [x] write tests

deweerdt commented on lib/handler/server_timing.c:
```diff
@@ -0,0 +1,53 @@
+    if (req->version == 0x200) {
+        /* ok */
+    } else if (req->version == 0x101 && req->res.content_length == SIZE_MAX) {
+        /* ok */
```
Should this check that the client sent `TE:trailers`?

kazuho commented:
> Should this check that the client sent `TE:trailers`?
> https://github.com/h2o/h2o/pull/1646#discussion_r167006241

@deweerdt Thank you for pointing that out. Addressed in cdb929d.

kazuho commented:
ToDo:
* [ ] redefine as a flag in `h2o_pathconf_t` instead of a filter, since the logic is tightly coupled to each protocol (and this includes moving chunked encoder into http1.c). Or can we (should we) define an interface for emitting trailers?
* [ ] tests

kazuho commented:
> redefine as a flag in `h2o_pathconf_t` instead of a filter, since the logic is tightly coupled to each protocol (and this includes moving chunked encoder into http1.c). Or can we (should we) define an interface for emitting trailers?

Or we can keep the code as-is; hoping that some day we will have a proper API. It's a bit annoying that the logic is spread into the filter and the protocol implementations, but practically speaking, the complexity would be about the same even if we refactor the design the way being proposed.

i110 commented:
@kazuho 
At https://github.com/h2o/h2o/pull/1646/commits/6006fbf1096c33dc9be6e257c85ef2301c937ccd I added more metrics that can be used in access log (connect-time, header-time, etc..)

The server-timing header value will be like:
```
server-timing: connect; dur=2.366, header; dur=0, body; dur=0, request_total; dur=0, process; dur=1004.742
(snip)
server-timing: response; dur=0, total; dur=1004.742
```

And I also added some tests at https://github.com/h2o/h2o/pull/1646/commits/8501e7a921f42b72d58080299c3b3beba1d705ce.

kazuho commented on lib/core/util.c:
```diff
@@ -769,6 +769,117 @@ h2o_iovec_t h2o_build_destination(h2o_req_t *req, const char *prefix, size_t pre
+    EMIT_ELEMENT(req, &dst, header, max_len);
+    EMIT_ELEMENT(req, &dst, body, max_len);
+    EMIT_ELEMENT(req, &dst, request_total, max_len);
+    EMIT_ELEMENT(req, &dst, process, max_len);
```
I believe that the names should match those defined for the Access Timings extensions defined for the access-log handler, with the `-time` suffix being omitted.

In other words, how about:
1. changing the names to be: connect, request-header, request-body, request-total, process, response
2. define `total-time` as a synonym of `duration` in the access-log handler

2 ensures us that we can continue to emit `server-timing: total; dur=xxx` as the trailer.

i110 replied:
Thank you for your comment. I addressed 1 at https://github.com/h2o/h2o/pull/1646/commits/51125dcc6cf206de18e3d5e24f1219f55fab71d0 and 2 at https://github.com/h2o/h2o/pull/1646/commits/3a8f262078203321163386685f68c72b47df4c77

kazuho commented on include/h2o.h:
```diff
@@ -2240,6 +2257,7 @@ COMPUTE_DURATION(request_total_time, &req->timestamps.request_begin_at, &req->pr
 COMPUTE_DURATION(process_time, &req->processed_at.at, &req->timestamps.response_start_at);
 COMPUTE_DURATION(response_time, &req->timestamps.response_start_at, &req->timestamps.response_end_at);
 COMPUTE_DURATION(duration, &req->timestamps.request_begin_at, &req->timestamps.response_end_at);
+COMPUTE_DURATION(total, &req->timestamps.request_begin_at, &req->timestamps.response_end_at);
```
I do not think that we need both `duration` and `total` as function definitions, since we do not need to guarantee source-level compatibility.

I would suggest either a) remove `duration` and rename `total` to `total_time` or b) remove `total`.

i110 replied:
Addressed at https://github.com/h2o/h2o/pull/1646/commits/cd6e57c135303667b149c199ba43d7e220bdc14a, but one thing I'm a bit wondering is about duration-stats, which still outputs only `duration` field. Which is the best choice?

1. Keep it as-is (total-time is not needed)
2. Add `total` (but it's redundant)
3. Remove `duration` and add `total` (of cause it may break backward compat)

kazuho commented on lib/core/logconf.c:
```diff
@@ -63,6 +63,7 @@ enum {
     ELEMENT_TYPE_PROCESS_TIME,                  /* %{process-time}x */
     ELEMENT_TYPE_RESPONSE_TIME,                 /* %{response-total-time}x */
     ELEMENT_TYPE_DURATION,                      /* %{duration}x */
+    ELEMENT_TYPE_TOTAL,                         /* %{total}x */
```
Same here. Both `%{duration}x` and `%{total-time}x` can map to the same enum element.

kazuho commented on lib/core/logconf.c:
```diff
@@ -246,6 +247,7 @@ h2o_logconf_t *h2o_logconf_compile(const char *fmt, int escape, char *errbuf)
                     MAP_EXT_TO_TYPE("process-time", ELEMENT_TYPE_PROCESS_TIME);
                     MAP_EXT_TO_TYPE("response-time", ELEMENT_TYPE_RESPONSE_TIME);
                     MAP_EXT_TO_TYPE("duration", ELEMENT_TYPE_DURATION);
+                    MAP_EXT_TO_TYPE("total", ELEMENT_TYPE_TOTAL);
```
The name should be `total-time`. `%{total}x` is too hard to understand. Note also that we have `-time` suffix for other elements.

kazuho commented on lib/core/util.c:
```diff
@@ -769,6 +769,115 @@ h2o_iovec_t h2o_build_destination(h2o_req_t *req, const char *prefix, size_t pre
+    usec -= ((int64_t)msec * 1000);
+    char *pos = buf;
+    pos += sprintf(pos, "dur=%" PRId32, msec);
+    if (usec != 0) {
```
Can't we simply use `sprintf(pos, "%dur=%" PRId32 ".%03u", msec, (unsigned)usec)` here?

I think that always expressing in the same precision is better in terms of consistency. Or is emitting `0`s at the end prohibited by the standard?

i110 replied:
It's not prohibited, but there is another care for the length of header value in https://www.w3.org/TR/server-timing/:

```
To minimize the HTTP overhead the provided names and descriptions should be kept as short as possible - e.g. use abbreviations and omit optional values where possible.
```

Similarly, I thought it's better to keep the duration value as short as possible. What do you think?

kazuho replied:
Thanks for elaborating. Let's keep the code as-is.
I can understand your point and I do not have a strong objection.

kazuho commented:
Thank you for the changes!

kazuho commented:
Note: this will be considered a "experimental" feature in 2.3. We might want to change how the feature is activated.
